### PR TITLE
GRAPHICS: Implement KotOR animations

### PR DESCRIPTION
**Pull request by seedhartha**
Mostly working KotOR animations implementation. Uses CPU for skeletal animation. There are some unhandled model node controller types (namely, compressed quaternions).

### CODE REVIEW

**Comment by DrMcCoy** [comment]
Hej, just wanted to say that I haven't forgotten your PR. I haven't really had enough time at a stretch to properly look at it, sorry.

Still, thank you for your work here! :)

File: src/graphics/aurora/modelnode.cpp [hunk 1]
```diff
@@ -162,13 +182,17 @@ void ModelNode::getAbsolutePosition(float &x, float &y, float &z) const {
 	z = _absolutePosition.getZ() * _model->_scale[2];
 }
 
-Common::Matrix4x4 ModelNode::getAsolutePosition() const {
+Common::Matrix4x4 ModelNode::getAbsolutePosition() const {
```
Comment (farmboy0) [inline src/graphics/aurora/modelnode.cpp]: This should be in a seperate commit

File: src/graphics/aurora/modelnode.h [hunk 2]
```diff
@@ -202,6 +218,8 @@ class ModelNode {
 	Common::BoundingBox _boundBox;
 	Common::BoundingBox _absoluteBoundBox;
 
+	uint16 _nodeNumber;
```
Comment (farmboy0) [inline src/graphics/aurora/modelnode.h]: _nodeNumber isnt initialized in the constructor.

File: src/graphics/aurora/model_kotor.cpp [hunk 3]
```diff
@@ -650,6 +744,51 @@ void ModelNode_KotOR::readMesh(Model_KotOR::ParserContext &ctx) {
...
+	_mesh->skin = new Skin();
+	_mesh->skin->boneMapping = new float[boneMappingCount];
+	_mesh->skin->boneMappingCount = boneMappingCount;
+	_mesh->skin->boneWeights = new float[_vertexCount * 4];
+	_mesh->skin->boneMappingId = new float[_vertexCount * 4];
+
+	uint32 pos = ctx.mdl->seek(ctx.offModelData + boneMappingOffset);
+	for (uint32 i = 0; i < boneMappingCount; i++)
+		_mesh->skin->boneMapping[i] = ctx.mdl->readIEEEFloatLE();
+	ctx.mdl->seek(pos);
```
Comment (farmboy0) [inline src/graphics/aurora/model_kotor.cpp]: This should jump to the end of the skin data but not only does it ignore the bone mappings it also ignores all the rest of the skin data whose definition you ignore in line 759.

File: src/graphics/aurora/model_kotor.h [hunk 4]
```diff
@@ -93,9 +95,14 @@ class ModelNode_KotOR : public ModelNode {
 	void load(Model_KotOR::ParserContext &ctx);
 
 private:
+	uint32 _mdxStructSize;
+	uint16 _vertexCount;
+	uint32 _offNodeData;
```
Comment (farmboy0) [inline src/graphics/aurora/model_kotor.h]: These arent used past the parsing of the model and move to the ParserContext

File: src/graphics/aurora/modelnode.h [hunk 5]
```diff
@@ -129,9 +130,20 @@ class ModelNode {
 		Dangly();
 	};
 
+	struct Skin {
+		float      *boneMapping;
+		uint32      boneMappingCount;
+		float      *boneWeights;
+		float      *boneMappingId;
+		ModelNode **boneNodeMap;
```
Comment (DrMcCoy) [inline src/graphics/aurora/modelnode.h]: Could you convert these arrays to `std::vector`s, please?

In general, I'd like to keep explicit new/delete to a minimum, by either using STL containers or ScopedPtr/ScopedArray.

File: src/graphics/aurora/model_kotor.cpp [hunk 6]
```diff
@@ -472,10 +506,64 @@ void ModelNode_KotOR::load(Model_KotOR::ParserContext &ctx) {
...
+		uint32 type = ctx.mdl->readUint32LE();
+		ctx.mdl->skip(2);
+		uint16 rowCount = ctx.mdl->readUint16LE();
+		uint16 timeIndex = ctx.mdl->readUint16LE();
+		uint16 dataIndex = ctx.mdl->readUint16LE();
+		uint8 columnCount = ctx.mdl->readByte();
+		ctx.mdl->skip(3);
+		switch (type) {
+			case kControllerTypePosition:
+				switch (columnCount) {
```
Comment (farmboy0) [inline src/graphics/aurora/model_kotor.cpp]: I think the inner switches should move to their own methods. Too much intentation here to make code easily readable.

File: src/graphics/aurora/model.h [hunk 7]
```diff
@@ -177,6 +177,8 @@ class Model : public GLContainer, public Renderable {
 	void render(RenderPass pass);
 	void advanceTime(float dt);
 
+	bool _isSkinned;
```
Comment (farmboy0) [inline src/graphics/aurora/model.h]: This flag shouldn't be public.

File: src/graphics/aurora/model_kotor.cpp [hunk 8]
```diff
@@ -387,6 +419,8 @@ void ModelNode_KotOR::load(Model_KotOR::ParserContext &ctx) {
 	uint16 superNode  = ctx.mdl->readUint16LE();
 	uint16 nodeNumber = ctx.mdl->readUint16LE();
 
+	_nodeNumber = nodeNumber;
```
Comment (farmboy0) [inline src/graphics/aurora/model_kotor.cpp]: This can be set and used directly

File: src/graphics/aurora/model_kotor.cpp (hunk 3 above)
Comment (farmboy0) [inline src/graphics/aurora/model_kotor.cpp]: This should go back to the position after the skin header(where you are after line 759) . You cant assume that the start of the bone mapping matches the end of the skin header.

File: src/graphics/aurora/modelnode.cpp (hunk 1 above)
Comment (seedhartha) [reply src/graphics/aurora/modelnode.cpp]: Fixed

File: src/graphics/aurora/modelnode.h (hunk 2 above)
Comment (seedhartha) [reply src/graphics/aurora/modelnode.h]: Fixed

File: src/graphics/aurora/model_kotor.cpp (hunk 3 above)
Comment (seedhartha) [reply src/graphics/aurora/model_kotor.cpp]: Not sure I understand you there. It doesn't read bone data from the end of the skin header. First it moves to the offset specified in skin header, reads the data, then returns to previous position.

File: src/graphics/aurora/model_kotor.cpp (hunk 8 above)
Comment (seedhartha) [reply src/graphics/aurora/model_kotor.cpp]: Fixed

File: src/graphics/aurora/model.h (hunk 7 above)
Comment (seedhartha) [reply src/graphics/aurora/model.h]: Made it protected

File: src/graphics/aurora/model_kotor.cpp (hunk 6 above)
Comment (seedhartha) [reply src/graphics/aurora/model_kotor.cpp]: Fixed

File: src/graphics/aurora/modelnode.h (hunk 5 above)
Comment (seedhartha) [reply src/graphics/aurora/modelnode.h]: Fixed

File: src/graphics/aurora/model_kotor.h (hunk 4 above)
Comment (seedhartha) [reply src/graphics/aurora/model_kotor.h]: Fixed

File: src/graphics/aurora/model_kotor.cpp (hunk 3 above)
Comment (farmboy0) [reply src/graphics/aurora/model_kotor.cpp]: The previous position you go back to is the start of the bone data.
The previous position you should go back is the end of the skin header.

Comment (seedhartha) [reply src/graphics/aurora/model_kotor.cpp]: According to the seek method description and your own binary template I'm pretty sure it goes back to the end of skin header as is.

Comment (farmboy0) [reply src/graphics/aurora/model_kotor.cpp]: You're right. I didn't realize seek returns the previous position before the seek.

**Comment by seedhartha** [comment]
All requested changes made. I had however accidentally pushed animation testing code if that's a problem.

**Comment by farmboy0** [comment]
Theres 2 things left to do:
1. use std::vector<float> instead of the float arrays or float * as DrMcCoy asked
2. the debugging code needs to go

**Comment by seedhartha** [comment]
Okay, that should be it.

**Comment by DrMcCoy** [comment]
Okay, merged it, thanks! :)

**Pull request merged**